openfermioncirq/primitives/_ffft.py

```diff
@@ -0,0 +1,192 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""The fast fermionic Fourier transform."""
+
+from typing import (List, Sequence)
+
+import numpy as np
+
+import cirq
+import cirq.contrib.acquaintance.permutation
+from openfermioncirq import FSWAP
+
+
+class _F0Gate(cirq.TwoQubitMatrixGate):
+    r"""Two-qubit gate that performs fermionic Fourier transform of size 2.
+
+    Realizes unitary gate :math:`F_0` that transforms Fermionic creation
+    operators :math:`a_0^\dagger` and :math:`a_1^\dagger` according to:
+
+    .. math::
+        F_0^\dagger a_0^\dagger F_0 =
+            {1 \over \sqrt{2}} (a_0^\dagger + a_1^\dagger)
+
+    .. math::
+        F_0^\dagger a_1^\dagger F_0 =
+            {1 \over \sqrt{2}} (a_0^\dagger - a_1^\dagger) \, .
+
+    This gate assumes JWT representation of fermionic modes which are big-endian
+    encoded on consecutive qubits:
+    :math:`a_0^\dagger \lvert 0 \rangle = \lvert 10_2 \rangle` and
+    :math:`a_1^\dagger \lvert 0 \rangle = \vert 01_2 \rangle`.
+
+    Internally, this leads to expansion of :math:`F_0^\dagger`:
+
+    .. math::
+        \langle 0 \rvert F_0^\dagger \lvert 0 \rangle = 1
+
+    .. math::
+        \langle 01_2 \rvert F_0^\dagger \lvert 01_2 \rangle =
+            -{1 \over \sqrt{2}}
+
+    .. math::
+        \langle 10_2 \rvert F_0^\dagger \lvert 10_2 \rangle =
+        \langle 10_2 \rvert F_0^\dagger \lvert 01_2 \rangle =
+        \langle 01_2 \rvert F_0^\dagger \lvert 10_2 \rangle = {1 \over \sqrt{2}}
+
+    .. math::
+        \langle 11_2 \rvert F_0^\dagger \lvert 11_2 \rangle = -1 \, .
+    """
+
+    def __init__(self):
+        """Initializes :math:`F_0` gate."""
+        cirq.TwoQubitMatrixGate.__init__(
+            self,
+            np.array([[1,          0,         0,  0],
+                      [0, -2**(-0.5), 2**(-0.5),  0],
+                      [0,  2**(-0.5), 2**(-0.5),  0],
+                      [0,          0,         0, -1]]))
+
+    def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs
+                               ) -> cirq.CircuitDiagramInfo:
+        if args.use_unicode_characters:
+            symbols = 'F₀', 'F₀'
+        else:
+            symbols = 'F0', 'F0'
+        return cirq.CircuitDiagramInfo(wire_symbols=symbols)
+
+
+class _TwiddleGate(cirq.ZPowGate):
```

Comment:
Does this need to inherit from ZPowGate? It could just specify that gate as its decomposition.

Reply:
Yea mean to inherit from `EigenGate`? I see that I would need to implement `__str__` and `__repr__` but I most likely should do that anyway. Except for `_eigen_components`, what am I require to support?

Reply:
Thanks, done!